### PR TITLE
perceive #26 Python notebook for capec visualization and images

### PR DESCRIPTION
**Pull request by vigneshrajan94**
added capec visualization notebook and images used

### CODE REVIEW

**Comment by vigneshrajan94** [comment]
[Vignesh - Week 5/12 - Python Notebook](https://trello.com/c/lbwKbizS/101-vignesh-week-5-12-python-notebook)

**Review by carlosparadis — CHANGES_REQUESTED** [review]
The notebook is top-notch! Really well done Vignesh :-) I have a few suggestions for the wording, in order to increase clarity for someone not used to XML. 

1) As I explained in that 17 minutes video, try to use table of tables analogies rather than tags. You should still mention "tag" but in parenthesis so someone who knows XML can also understand. I would also like you to introduce a new header just before this passage:

> The Categories are represented in the Categories tag of the XML and their titles are mentioned in attributes. Similarly, the Meta attack patterns are represented in the Attack Patterns tag of the XML and their titles are mentioned in attributes. 

The header should contain the words "Data Model". Since this is really what you are getting into here. Showing the XML is fine, but it is too overwhelming for someone not acquainted with it. Please take a few screenshots of that video I sent to you, showing the same XML information as the tables of tables visualization from the software I was using, and add here. Even better, download the tool and take one yourself. Add one for every pair of XML figure you have here. Additionally, the first thing you should mention before you delve into the details of the Categories table is to repeat the name of the 4 tables again (trust me, with this amount of information it is very fast to forget). 

- Provide an example of what you mean by "shown in the tag" or "shown in the attribute" immediately after you mention them with a figure.

- Since your notebook is not an introduction like the other one, you should make your statement purpose even before your own introduction, and link to the other capec notebook on it as a starting point. 

- Your script doesn't need to be inside a single block. Break it into the logical parts so people can see the intermediate results. 

> The first part of the program creates a CSV file which contains CAPEC_ID, CAPEC_NAME, CHILD_OF columns and parses the XML to populate various categories, meta attack patterns, and their child standard attack patterns.

Put the first block of code here associated with this and some intermediate outputs. 

> The second part involves parsing this CSV file to a JSON object that can be used for the Foamtree visualization. The JSON is then fed into the dataObject variable in the JavaScript and hence must follow the standards set by Foamtree.

Same here as above. 

> The final part involves putting together the JavaScript file whose path has been obtained as input 2 in an HTML file and then adding the JSON data into the dataObject variable in the JavaScript to obtain the final visualization.

And same as above. 

We can make a separate py file that is self-contained with everything if it needs to be. 

- Your indentation is too big between the lines, it's even impossible to see on GitHub without the text appearing on the other line wrapped. Try to reduce it! Usually, 4 paces are enough to distinguish one line for the other. 

I will provide more feedback on your code once you split into the 3 blocks! 

Good job! 👍

**Comment by vigneshrajan94** [comment]
I am making the changes according to the feedback you have given. 

However, I am having issues with one of them. Using Liquid XML to view the "Related Attack Patterns" of a Meta Attack Pattern is very difficult in it as the various tags are represented from left to right and not one below the other as here:

The point is to show that Meta Attack Patterns do not have this "Related Attack Patterns" while Standard Attack Patterns do. 

This is the one I used
image

This is using Liquid XML:
image

You can see that the tag is nowhere in the image.

Should I still use this tool or should I stick with the one I used?

**Comment by carlosparadis** [comment]
@vigneshrajan94 Both. Not everything is clear in one tool or the other. For this case where the thing just keeps going on forever and can't fit the screen, just mention this reflects on the column missing. That is perfectly fine. I think for the argument of "consistency" of a given information, Liquid XML is superior, since it does fixate a column with grayed out and filled information; or to show how hard certain information it is to get out. Makes sense? It is best to use consistently on both than using either when convenient, as this would make it hard for someone using one of the 2 tools.

**Comment by vigneshrajan94** [comment]
@carlosviansi Regarding the indentation, I have used tab space which is equal to four spaces. I am unable to modify that without throwing errors.

**Comment by vigneshrajan94** [comment]
@carlosviansi Other than the indentation issue, I have modified the rest of the notebook. I will make further changes based on the comments. Should I create a new pull request or modify this current one?

**Comment by vigneshrajan94** [comment]
@carlosviansi I've replaced the tab indent with four spaces and I think that should deal with the indentation problem. Also I've deleted the initial notebook file and uploaded the modified file. Kindly review and provide me with feedback :)

**Comment by carlosparadis** [comment]
@vigneshrajan94 Thanks! It looks good! I will take a more in-depth pass tomorrow and let you know!

**Comment by vigneshrajan94** [comment]
@carlosviansi Hello there :)

**Comment by carlosparadis** [comment]
@vigneshrajan94 Hey! Thanks for checking this. I gave you a follow-up for this week-todo. It doesn't involve updating this notebook per se, so it should be safe for me to push this notebook review a bit further down the road :-) Thanks for the reminder, they do help a lot!

**Review by carlosparadis — CHANGES_REQUESTED** [review]
Great work. I have some requests on your code still. I see it still expects a .csv format to generate a .json format. I believe we can simplify the logic further. 

Your code should have more-or-less these logic blocks:

- Parsing the XML into memory, into a **data model** that can retain the structure in order to generate the following models. This would be trivial, if not by the fact CAPEC data contains the information we want to be split into 3 tables. Specifically, the information is the Mechanisms of Attack hierarchical relationship, **ignoring** the remaining relationships any one entry has. 

 - When represented in memory, this should be unified under a combination of python lists and dictionaries, that reflects our understanding in how CAPEC should have been apriori and in fact, as it is shown on the website! We have a `root node` which is the Mechanism of Attack View,  which contains a python dictionary of categories, indexed by their id and name key, whose values are another dictionary containing the remaining hierarchy. Provided this **unified memory representation, which reflects that of the website and is closer to intuition**, then we should have "export functions". 

- One is the .csv you already have, which in reality is an `export_edgelist` function, as it is the precise format one would load in Gephi. 

- The second is that of the foamtree. 

- The third could be something else. 

As you can see, the model is more flexible if in memory you represent this unified representation, and the functions then defer the appropriate format. In fact, I imagine the foamtree closely expects the model I am suggesting here.

There will be another huge benefit in removing these unnecessary .csv dependencies: We can parameterize your script to, provided an --edgelist, or --json, or both it can export them independently. 

In summary, the functionality is the same, but the logic needs to be refactored. Also, please use functions to convey the logic. The comments can be moved to function documentation (''' 3 quotes ''' under function definition). 

Minor fix: at the bottom, add the name of the CAPEC IDS you are referring. The numbers by themselves, other than an expert are impossible to understand without googling each one of them.

**Comment by carlosparadis** [comment]
Also, please note that it is of our interest to later expand to other Views, hence, if we have this representation, extending to another View is merely adding more root nodes to it for every view. I am also hoping we can re-run this code for CWE, whose format is very similar in terms of logic. Will discuss more about this once I set next week cards.

**Comment by vigneshrajan94** [comment]
@carlosviansi The changes sound great and I will start working on them right away. I have a few questions though, 

1. When you say "Parsing the XML into memory, into a data model that can retain the structure in order to generate the following models.", do you mean the information that is related to Mechanisms of Attack only? 

2. So I am guessing I will create a data model which has various nodes one for each view - one for Mechanisms of attack and so on. By default, I will direct it to the Mechanisms of Attack view and this can be modified easily. I will create the combination of lists and dictionaries to list out all the attack patterns under that view and their details bundled in as dictionaries. Have I got this right or should I do anything differently?

Also I will make the following changes:

1. Remove the dependency on CSV.
2. Build the export functions.
3. Improve function documentation.
4. Including those CAPEC IDS and other minor text fixes.

Will begin as soon as you approve of my understanding.

**Comment by carlosparadis** [comment]
@vigneshrajan94 sorry, 19 days ago I won't have any more trace or memory of what I said back then. Avoiding long breaks like these is beneficial for you in having me remember what is going on as well. I am going to send you a voice audio but I believe your understanding is correct. It should be doable for your in-memory data model to load other views (Whether from CAPEC or CWE). Please also remember you need to investigate the more detailed foamtree representation.

**Comment by vigneshrajan94** [comment]
I am still waiting for the voice audio. And I will investigate the more detailed foamtree representation.

**Comment by carlosparadis** [comment]
@vigneshrajan94 I sent that e-mail to replace the voice audio, do you still have questions associated to this or to the tasks there? If yes, maybe we should do a quick skype call. Let me know!

**Comment by vigneshrajan94** [comment]
@carlosviansi I read that email but I still have some questions. Let me know when you will be available for a call today.

**Comment by vigneshrajan94** [comment]
I have added the final capec_visualisation notebook and also associated file. It contains all the changes that were requested.

**Comment by vigneshrajan94** [comment]
I am sorry I need to add a thorough discussion of what I have learnt from the visualisation and from the code and also provide my recommendations for the corpus.

**Comment by vigneshrajan94** [comment]
I have updated the notebook with all the required information. Hope it helps.

**Comment by vigneshrajan94** [comment]
I've added the difference report notebook which contains details as to how to build the CSV and the associated visualization.

**Comment by carlosparadis** [comment]
@vigneshrajan94 There is a conflict on capec_introduction.ipynb, I can't accept the merge as it is. Could you solve it?

**Comment by carlosparadis** [comment]
For future reference: This pull request was sent in separate parts and already accepted. It should not be re-opened.
